examples/petsc/multigrid.c

```diff
@@ -356,56 +345,40 @@ int main(int argc, char **argv) {
-
-    // -- Local diagonal
-    CeedOperatorAssembleLinearDiagonal(userO[i]->op, &diagceed,
-                                       CEED_REQUEST_IMMEDIATE);
+    if (i > 0) {
+      // Prolongation/Restriction Operator
+      userPR[i]->comm = comm;
+      userPR[i]->dmF = dm[i];
+      userPR[i]->dmC = dm[i-1];
+      userPR[i]->locVecC = Xloc[i-1];
+      userPR[i]->locVecF = userO[i]->Yloc;
+      userPR[i]->multVec = mult[i];
+      userPR[i]->ceedVecC = userO[i-1]->xceed;
+      userPR[i]->ceedVecF = userO[i]->yceed;
+      userPR[i]->opProlong = ceeddata[i]->opProlong;
+      userPR[i]->opRestrict = ceeddata[i]->opRestrict;
+      userPR[i]->ceed = ceed;
+    }
+  }
 
-    // -- Set PETSc array
-    CeedVectorGetArrayRead(diagceed, CEED_MEM_HOST, &ceedarray);
-    ierr = VecPlaceArray(Xloc[i], ceedarray); CHKERRQ(ierr);
-    CeedVectorRestoreArrayRead(diagceed, &ceedarray);
+  // Setup dummy SNES for AMG coarse solve
+  ierr = SNESCreate(comm, &snes_dummy); CHKERRQ(ierr);
+  ierr = SNESSetDM(snes_dummy, dm[0]); CHKERRQ(ierr);
+  ierr = SNESSetSolution(snes_dummy, X[0]); CHKERRQ(ierr);
 
-    // -- Global diagonal
-    ierr = VecZeroEntries(userO[i]->diag); CHKERRQ(ierr);
-    ierr = DMLocalToGlobalBegin(userO[i]->dm, Xloc[i], ADD_VALUES,
-                                userO[i]->diag); CHKERRQ(ierr);
-    ierr = DMLocalToGlobalEnd(userO[i]->dm, Xloc[i], ADD_VALUES,
-                              userO[i]->diag); CHKERRQ(ierr);
+  // -- Jacobian matrix
+  ierr = DMSetMatType(dm[0], MATAIJ); CHKERRQ(ierr);
+  ierr = DMCreateMatrix(dm[0], &matCoarse); CHKERRQ(ierr);
+  ierr = SNESSetJacobian(snes_dummy, matCoarse, matCoarse, NULL,
+                         NULL); CHKERRQ(ierr);
 
-    // -- Cleanup
-    ierr = VecResetArray(Xloc[i]); CHKERRQ(ierr);
-    CeedVectorDestroy(&diagceed);
+  // -- Residual evaluation function
+  ierr = SNESSetFunction(snes_dummy, X[0], FormResidual_Ceed,
+                         userO[0]); CHKERRQ(ierr);
 
-    if (i > 0) {
-      // Interp Operator
-      userI[i]->comm = comm;
-      userI[i]->dmc = dm[i-1];
-      userI[i]->dmf = dm[i];
-      userI[i]->Xloc = Xloc[i-1];
-      userI[i]->Yloc = userO[i]->Yloc;
-      userI[i]->mult = mult[i];
-      userI[i]->ceedvecc = userO[i-1]->xceed;
-      userI[i]->ceedvecf = userO[i]->yceed;
-      userI[i]->op = ceeddata[i]->op_interp;
-      userI[i]->ceed = ceed;
-
-      // Restrict Operator
-      userR[i]->comm = comm;
-      userR[i]->dmc = dm[i-1];
-      userR[i]->dmf = dm[i];
-      userR[i]->Xloc = Xloc[i];
-      userR[i]->Yloc = userO[i-1]->Yloc;
-      userR[i]->mult = mult[i];
-      userR[i]->ceedvecf = userO[i]->xceed;
-      userR[i]->ceedvecc = userO[i-1]->yceed;
-      userR[i]->op = ceeddata[i]->op_restrict;
-      userR[i]->ceed = ceed;
-    }
-  }
+  // -- Form Jacobian
+  ierr = SNESComputeJacobianDefaultColor(snes_dummy, X[0], matO[0],
+                                         matCoarse, NULL); CHKERRQ(ierr);
```

Comment:
You don't have to make a dummy SNES to use coloring, though I suppose you could if your preference is to maintain consistency with the hyperelasticity code. The interface you'd use is basically as in `SNESComputeJacobianDefaultColor`:
```c
      ierr = DMCreateColoring(dm,IS_COLORING_GLOBAL,&iscoloring);CHKERRQ(ierr);
      ierr = MatFDColoringCreate(B,iscoloring,&color);CHKERRQ(ierr);
      ierr = MatFDColoringSetFunction(color,(PetscErrorCode (*)(void))SNESComputeFunctionCtx,NULL);CHKERRQ(ierr);
      ierr = MatFDColoringSetFromOptions(color);CHKERRQ(ierr);
      ierr = MatFDColoringSetUp(B,iscoloring,color);CHKERRQ(ierr);
      ierr = ISColoringDestroy(&iscoloring);CHKERRQ(ierr);
```

Reply:
Yes, my thought was to make a simplified example of libCEED matrix assembly via SNES that we see in the elasticity code.

Reply:
Cool. We could use SNES even for the top-level linear problem. (I've often done that when writing linear examples myself.)